relay-sampling/src/lib.rs

```diff
@@ -759,9 +766,24 @@ impl DynamicSamplingContext {
     /// configuration is invalid.
     pub fn should_keep(&self, ip_addr: Option<IpAddr>, config: &SamplingConfig) -> SamplingResult {
         if let Some(rule) = get_matching_trace_rule(config, self, ip_addr, RuleType::Trace) {
+            let client_sample_rate = self.sample_rate.map(|x| x.0).unwrap_or(1.0);
+            let mut adjusted_sample_rate = (rule.sample_rate / client_sample_rate).clamp(0.0, 1.0);
+
+            if adjusted_sample_rate.is_infinite() || adjusted_sample_rate.is_nan() {
+                // adjusted_sample_rate is infinite or NaN. This is a bug.
+                //
+                // - infinite should not happen because we clamped the number.
+                // - NaN can happen if the client_sample_rate is 0, which is bogus because the SDK
+                //   should've dropped the envelope. In that case let's pretend the sample rate was
+                //   not sent, because clearly the sampling decision across the trace is still 1.
+                //   The most likely explanation is that the SDK is reporting its own sample rate
+                //   setting instead of the one from the continued trace.
+                adjusted_sample_rate = 1.0;
+            }
```

Comment:
Even though the formula is simple, it would be nice to wrap this in a function `adjust_sample_rate()` with unit tests for a few cases (e.g. `(0,1), (1,1), (0.1, 0.5), (0.5, 0.1)` etc.).

Reply:
done